File: contract/swift/Sources/KernelModel/Receiver.swift
import Foundation

/// K6 Tenancy and policy (contract/spec/K6-tenancy-policy.md).
public struct Caller: Hashable, Codable, Sendable {
    public var tenantId: String
    public var principalId: String

    public init(tenantId: String, principalId: String) {
        self.tenantId = tenantId
        self.principalId = principalId
    }
}

/// Applies the kernel's receiving order (T2) for one authority.
public struct Receiver: Sendable {
    public var changes: ChangeLog
    public var authorities: Authorities
    /// Supplied by the domain (T3).
    public var policy: @Sendable (Caller, Submission) -> Bool

    public init(changes: ChangeLog, authorities: Authorities, policy: @escaping @Sendable (Caller, Submission) -> Bool) {
        self.changes = changes
        self.authorities = authorities
        self.policy = policy
    }

    /// Accepts or rejects `s`; `domain` runs the domain's rules (K4 C10).
    public mutating func receive(_ caller: Caller, _ s: Submission, at now: Date,
                                 domain: (() throws(KernelError) -> Void)? = nil) throws(KernelError) -> ChangeRecord {
        guard s.tenantId == caller.tenantId, s.principalId == caller.principalId else { throw .policyDenied } // T1
        let authorities = authorities, policy = policy
        return try changes.submit(s, at: now) { () throws(KernelError) in
            try authorities.authorize(s)                                   // A3
            guard policy(caller, s) else { throw .policyDenied }           // T3
            try domain?()
        }
    }
}

File: contract/go/kernel/receiver.go
package kernel

import (
	"time"

	pb "platformkernel/gen/platform/kernel/v1alpha1"
)

// Caller is the principal a receiver authenticated (K6).
type Caller struct{ Tenant, Principal string }

// Policy decides whether a caller may submit s; supplied by the domain (K6 T3).
type Policy func(caller Caller, s *pb.Submission) bool

// Receiver applies the kernel's receiving order (K6 T2) for one authority.
type Receiver struct {
	Changes     *ChangeLog
	Authorities *Authorities
	Policy      Policy
}

// Receive accepts or rejects a submission; domain runs the domain's rules (K4 C10).
func (r *Receiver) Receive(c Caller, s *pb.Submission, now time.Time, domain func() *Error) (*pb.ChangeRecord, *Error) {
	if s.GetTenantId() != c.Tenant || s.GetPrincipalId() != c.Principal {
		return nil, errorf(pb.ErrorCode_ERROR_CODE_POLICY_DENIED) // T1
	}
	return r.Changes.SubmitChecked(s, now, func() *Error {
		if err := r.Authorities.Authorize(s); err != nil {
			return err // A3
		}
		if r.Policy == nil || !r.Policy(c, s) {
			return errorf(pb.ErrorCode_ERROR_CODE_POLICY_DENIED) // T3
		}
		if domain != nil {
			return domain()
		}
		return nil
	})
}
